File: PGBooks/PGBooks/Sources/UI/Common/Cells/BookSimpleCell.m
//
//  BookSimpleCell.m
//  PGBooks
//
//  Created by ipagong on 13/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

#import "BookSimpleCell.h"
#import "PGBook.h"
#import "CustomViews.h"
#import "UIImageView+AFNetworking.h"

@interface BookSimpleCell ()

@property (nonatomic, strong) NSString * url;

@end

@implementation BookSimpleCell

+ (void)registOnTableView:(UITableView *)tableview
{
    [tableview registerNib:[UINib nibWithNibName:[BookSimpleCell identifier] bundle:nil]
    forCellReuseIdentifier:[BookSimpleCell identifier]];
}

+ (NSString *)identifier {
    return NSStringFromClass([BookSimpleCell class]);
}

- (void)awakeFromNib {
    [super awakeFromNib];
    self.coverView.contentMode = UIViewContentModeScaleToFill;
}

- (void)prepareForReuse {
    [super prepareForReuse];

    [self.coverView cancelImageDownloadTask];
    
    self.coverView.image        = nil;
    self.titleLabel.text        = nil;
    self.subtitleLabel.text     = nil;
    self.priceLabel.text        = nil;
    self.url                    = nil;
    self.numberLabel.text       = nil;
}

- (void)bindData:(PGBook *)book
{
    self.key = book.isbn13;
    
    self.titleLabel.text        = book.title;
    self.subtitleLabel.text     = book.subtitle;
    self.priceLabel.text        = book.price;
    self.url                    = book.url;
    self.numberLabel.text       = book.isbn13;
    
    [self loadIamge:book.image];
}

- (void)loadIamge:(NSString *)imageURL {
    if (!imageURL) { return; }
    
    NSURLRequest * url = [NSURLRequest requestWithURL:[NSURL URLWithString:imageURL]];
    
    __weak __typeof(self)weakSelf = self;
    [self.coverView setImageWithURLRequest:url
                          placeholderImage:nil
                                   success:^(NSURLRequest * _Nonnull request, NSHTTPURLResponse * _Nullable response, UIImage * _Nonnull image) {
                                       __strong __typeof(self)strongSelf = weakSelf;
                                       strongSelf.coverView.image = image;
                                       
                                       [strongSelf.coverView setNeedsLayout];
                                       [strongSelf.coverView layoutIfNeeded];
                                   } failure:nil];
}

- (IBAction)onClickButton:(id)sender {
    if (self.url == nil) { return; }
    
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:self.url]
                                       options:@{}
                             completionHandler:nil];
}

@end

File: PGBooks-swift/PGBooks-swift/Sources/UI/Common/Cells/BookSimpleCell.swift
//
//  BookSimpleCell.swift
//  PGBooks-swift
//
//  Created by ipagong on 17/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

import UIKit
import AlamofireImage

class BookSimpleCell: UITableViewCell, CellFactory {
    static let identifier = "BookSimpleCell"
    
    @IBOutlet weak var coverView        : UIImageView!
    @IBOutlet weak var priceLabel       : PaddingLabel!
    @IBOutlet weak var titleLabel       : UILabel!
    @IBOutlet weak var subtitleLabel    : UILabel!
    @IBOutlet weak var storeLinkButton  : UIButton!
    @IBOutlet weak var numberLabel      : PaddingLabel!

    var data:Book?
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.coverView.af_cancelImageRequest()
        self.coverView.image = nil
    }
    
    func bindData(value: Book?) {
        self.data = value
        
        self.titleLabel.text     = value?.title
        self.subtitleLabel.text  = value?.subtitle
        self.priceLabel.text     = value?.price
        self.numberLabel.text    = value?.isbn13
        
        if let url = try? value?.image.asURL() { self.coverView.af_setImage(withURL: url) }
    }
    
    @IBAction func onClickButton(_ sender:Any?) {
        guard let url = try? data?.url.asURL() else { return }
        
        UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }
    
}
